clash-prelude/src/Clash/Prelude/Mealy.hs

```diff
@@ -129,6 +172,54 @@ mealyB
 mealyB = hideClockResetEnable E.mealyB
 {-# INLINE mealyB #-}
 
+
+-- | Create a synchronous function from a combinational function describing
+-- a mealy machine using the state monad. This can be particularly useful
+-- when combined with lenses or optics to replicate imperative algorithms.
+--
+-- @
+-- data DelayState = DelayState
+--   { _history    :: Vec 4 Int
+--   , _untilValid :: Index 4
+--   }
+-- makeLenses 'DelayedState
+--
+-- initialDelayState = DelayState (repeat 0) maxBound
+--
+-- delayS :: Int -> State DelayState (Maybe Int)
+-- delayS n = do
+--   history   %= (n +>>)
+--   remaining <- use untilValid
+--   if remaining > 0
+--   then do
+--      remaining -= 1
+--      return Nothing
+--    else do
+--      out <- uses history last
+--      return (Just out)
+--
+-- delayTop :: HiddenClockResetEnable dom  => 'Signal' dom Int -> 'Signal' dom (Maybe Int)
+-- delayTop = 'mealyS' delayS initialDelayState
+-- @
```

Comment:
```suggestion
-- @
-- data DelayState = DelayState
--   { _history    :: Vec 4 Int
--   , _untilValid :: Index 4
--   }
--   deriving (Generic, NFDataX)
-- makeLenses ''DelayState
--
-- initialDelayState = DelayState (repeat 0) maxBound
--
-- delayS :: Int -> State DelayState (Maybe Int)
-- delayS n = do
--   history   %= (n +>>)
--   remaining <- use untilValid
--   if remaining > 0
--   then do
--      untilValid -= 1
--      return Nothing
--    else do
--      out <- uses history last
--      return (Just out)
--
-- delayTop :: HiddenClockResetEnable dom  => 'Signal' dom Int -> 'Signal' dom (Maybe Int)
-- delayTop = 'mealyS' delayS initialDelayState
-- @
```

Reply:
Because GitHub doesn't really show it, I'll point out what this actually boils down to:
```diff
diff --git a/clash-prelude/src/Clash/Explicit/Mealy.hs b/clash-prelude/src/Clash/Explicit/Mealy.hs
index 08b23ca89..022a9e43f 100644
--- a/clash-prelude/src/Clash/Explicit/Mealy.hs
+++ b/clash-prelude/src/Clash/Explicit/Mealy.hs
@@ -159,7 +159,8 @@ mealy clk rst en f iS =
 --   { _history    :: Vec 4 Int
 --   , _untilValid :: Index 4
 --   }
--- makeLenses 'DelayedState
+--   deriving (Generic, NFDataX)
+-- makeLenses ''DelayState
 --
 -- initialDelayState = DelayState (repeat 0) maxBound
 --
@@ -169,7 +170,7 @@ mealy clk rst en f iS =
 --   remaining <- use untilValid
 --   if remaining > 0
 --   then do
---      remaining -= 1
+--      untilValid -= 1
 --      return Nothing
 --    else do
 --      out <- uses history last
```
and then the same for the other file.